src/approaches/grammar_search_invention_approach.py

```diff
@@ -373,6 +381,42 @@ def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
             yield (predicate, cost)
 
 
+@dataclass(frozen=True, eq=False, repr=False)
+class _PrunedGrammar(_DataBasedPredicateGrammar):
+    """A grammar that prunes redundant predicates.
+    """
+    base_grammar: _PredicateGrammar
+
+    def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
+        # Predicates are identified based on their evaluation across
+        # all states in the dataset.
+        seen = {}  # maps identifier to previous predicate, for debugging
+        for (predicate, cost) in self.base_grammar.enumerate():
+            if cost >= CFG.grammar_search_predicate_cost_upper_bound:
+                return
+            pred_id = self._get_predicate_identifier(predicate)
+            if pred_id in seen:
+                # Useful for debugging
+                # print("Pruning", predicate, "b/c equal to", seen[pred_id])
+                continue
+            # Found a new predicate.
+            seen[pred_id] = predicate
+            yield (predicate, cost)
+
+    def _get_predicate_identifier(self, predicate: Predicate) -> int:
+        """Returns a hash of frozensets of groundatoms for each data point.
+        """
+        # Get atoms for this predicate alone on the dataset.
+        atom_dataset = utils.create_ground_atom_dataset(self.dataset,
```

Comment:
lol, there's no way this hacky hashing works... does it really!?!

Reply:
wait what's hacky about it? the reason i'm storing `hash(thing)` rather than `thing` is because the `thing`s could be large, but I can probably just store the things anyway if this is confusing or seems bad

Reply:
eh nvm I'll just do the other way